{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c31db202-3b3c-486d-bf01-23cfce0b3b91",
   "metadata": {},
   "source": [
    "# 2. nfcore/rnaseq workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "775a59ed-32e4-4743-88bf-7dff19dcf4d6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "194c0c15-a083-490b-90f0-91d469be7623",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e4539ef-e293-4256-abdd-fbceb16301ab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "72868086-8902-4015-9722-5c5f4e3530b2",
   "metadata": {},
   "source": [
    "**********************************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28b6582d-4fbf-47f3-8d58-c24aa1b2c3c5",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Contents\n",
    "\n",
    "[2a. Prerequisites](#prereq2)\n",
    "\n",
    "[1b. Differential expression workflow overview](#overview2)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6c5a86c-b325-4a8e-9978-7f7bd9648213",
   "metadata": {},
   "source": [
    "***************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90b1faea-5c7f-4101-b2f9-c246c8a34ed1",
   "metadata": {},
   "source": [
    "## 2a. Prerequisites <a class=\"anchor\" id=\"prereq2\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24fed7cb-4a13-483b-9619-b8f0f2bfd5f1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "761338b7-44ce-49cb-bdc9-e121df1b7d84",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "44841dae-58a0-4388-9553-90bdec0b89aa",
   "metadata": {},
   "source": [
    "*****************"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e91926cb-8bca-4c0e-b1ca-9de08fec87d7",
   "metadata": {},
   "source": [
    "## 2b. Differential expression workflow overview <a class=\"anchor\" id=\"overview2\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e9e162c-0fbf-45b2-9807-0c284458a66c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "ae4a1c3b-5ff1-4cc4-a834-a979e6feeb8d",
   "metadata": {},
   "source": [
    "### Downstream: R packages"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9fafab0-c733-48c2-b9f3-faad93ba7c3a",
   "metadata": {},
   "source": [
    "A count table of reads per sample per gene (samples as columns, genes as rows) is generated by nfcore/rnaseq and is used as input for the downstram analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c089caf8-4755-4f95-8db3-ec676ba6ee60",
   "metadata": {},
   "source": [
    "The primary downstream analysis is differential expression, using the R package [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html):\n",
    "\n",
    "> Estimate variance-mean dependence in count data from high-throughput sequencing assays and test for differential expression based on a model using the negative binomial distribution\n",
    "\n",
    "From DESEq2 results, volcano plots, heapmaps and tables of significantly differentially expressed (DE) genes are genereated."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a762db0-84ee-4be6-a830-90452e03f788",
   "metadata": {},
   "source": [
    "Other downstream analyses include:\n",
    "\n",
    "* Outlier and batch effects\n",
    "\n",
    "  * **Density plots** show the distribution of read density per sample. They can identify whether particular samples have a bias in the read count per gene, per sample. A strong bias can be an indication of errors in library preparation and batch effects.\n",
    "  * [**PCA plots**](https://setosa.io/ev/principal-component-analysis/) show linear combinations of the ‘principal components’ in a dataset, i.e. the linearly uncorrelated variables as estimated by the R package `prcomp`. PC1 displays the majority of the variation, decreasing rapidly to PC2 and through to PCn (n = variable/sample size). Groups have been coloured and shaded based on 85% confidence levels for data range.\n",
    "  * [**Dendrograms and heatmaps**](https://docs.tibco.com/pub/spotfire/7.0.1/doc/html/heat/heat_what_is_a_heat_map.htm). Dendrograms show the hierarchical clustering of samples, based on pairwise Euclidean distance. Heatmaps show, using a colour gradient, pairwise Euclidean distance between individual samples.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e65fd352-9ee3-4c9d-ba3e-9870e6f30527",
   "metadata": {},
   "source": [
    "* Functional annotation (run in separate Notebook)\n",
    "\n",
    "  * [KEGG pathway](https://www.genome.jp/kegg/) and [GO term](http://geneontology.org/) enrichment, based on differentially expressed genes, using the R tool [clusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R .conda-rnotebook",
   "language": "R",
   "name": "conda-env-.conda-rnotebook-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
